### DISCUSSION TITLE
with_loader_criteria does not work with many to many relationship

### DISCUSSION THREAD

**Discussion by yanivps** [sqlalchemy/sqlalchemy#10250] [Usage Questions] [answered]
### Describe the bug

The `with_loader_criteria` does not seem to work with many to many relationship
I've made a small example which shows the problem and helps to reproduce it

In this use case I am using the `with_loader_criteria` in order to filter out records with `deleted_at` values not equal to NULL. It seems to work in one to many relationship but not in many to many relationships

Here is the full example:
```python
from typing import List
from sqlalchemy import Column, ForeignKey, Integer, String, create_engine
from sqlalchemy import event, orm
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, configure_mappers, relationship

engine = create_engine("postgresql://postgres:secret@localhost:5432/test_is_deleted", echo=True)
session = Session(engine)

Base = declarative_base()

from datetime import datetime
import sqlalchemy as sa
from sqlalchemy.orm import declarative_mixin


@declarative_mixin
class SoftDeleteable:
    """Mixin that identifies a class as having deleted_at
    column instead of being hard deleted from the database"""

    deleted_at = sa.Column(sa.DateTime(timezone=True), nullable=True, index=True)


class X(Base):
    __tablename__ = "x"
    id = Column(Integer, primary_key=True)
    ys: List["Y"] = relationship("Y", secondary="xy", uselist=True)
    xys: List["XY"] = relationship("XY", uselist=True)


class Y(Base):
    __tablename__ = "y"
    id = Column(Integer, primary_key=True)
    value = Column(String)

    def __repr__(self) -> str:
        return "Y(value={})".format(self.value)


class XY(SoftDeleteable, Base):
    __tablename__ = "xy"
    x_id = Column(Integer, ForeignKey(X.id), primary_key=True)
    y_id = Column(Integer, ForeignKey(Y.id), primary_key=True)

    def __repr__(self) -> str:
        return "XY(x_id={}, y_id={}, deleted_at={})".format(self.x_id, self.y_id, self.deleted_at)


@event.listens_for(Session, "do_orm_execute")
def _add_filtering_criteria(execute_state):  # type:ignore
    if (
        not execute_state.is_column_load
        and not execute_state.is_relationship_load
        and not execute_state.execution_options.get("include_deleted", False)
    ):
        execute_state.statement = execute_state.statement.options(
            orm.with_loader_criteria(
                SoftDeleteable,
                lambda cls: cls.deleted_at.is_(None),
                include_aliases=True,
            )
        )


configure_mappers()
Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

session.add(X(ys=[Y(value="Value 1"), Y(value="Value 2"), Y(value="Value 3")]))
session.commit()
some_xy = session.query(XY).join(Y).filter(Y.value == "Value 2").first()
some_xy.deleted_at = datetime.now()
session.commit()

# TEST STARTS HERE
some_x = session.query(X).first()

print(some_x.xys)
# The line above prints: [XY(x_id=1, y_id=1, deleted_at=None), XY(x_id=1, y_id=3, deleted_at=None)]
# which is good as the XY record with the non-null deleted_at value did not return
# Query:
# SELECT xy.deleted_at AS xy_deleted_at, xy.x_id AS xy_x_id, xy.y_id AS xy_y_id
# FROM xy
# WHERE 1 = xy.x_id AND xy.deleted_at IS NULL

print(some_x.ys)
# The line above prints: [Y(value=Value 1), Y(value=Value 2), Y(value=Value 3)]
# which is not good as the Y record with value="Value 2" should not have been returned
# because it is related to X via a "soft deleted" XY record
# Query:
# SELECT y.id AS y_id, y.value AS y_value
# FROM y, xy
# WHERE 1 = xy.x_id AND y.id = xy.y_id
```

It absolutely does not make any sense to fetch all Y records associated with X in this case as to get the Y records we need to go through a joining table XY where the `deleted_at` column must be respected via the `orm.with_loader_criteria`

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/14/_modules/examples/extending_query/filter_public.html

### SQLAlchemy Version in Use

1.4.42

### DBAPI (i.e. the database driver)

psycopg2

### Database Vendor and Major Version

PostgreSQL 14

### Python Version

3.8.10

### Operating system

OSX

### Error

No error

**Comment by CaselIT** [comment]
Hi,

So I think this is the intended behavior since when loading `ys` no `xy` is actually loaded.
I'm sure mike can provide additional info

**Comment by zzzeek** [comment]
hi -

The X.ys relationship is not related to the XY mapping in any way.  it happens to use the same underlying table as XY does for its `secondary` argument but that does not indicate any ORM instructions being shared between these two contexts.

if you want X.ys to load in terms of X.xys use the [association proxy](https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html)

**Comment by yanivps** [comment]
> hi -
> 
> The X.ys relationship is not related to the XY mapping in any way. it happens to use the same underlying table as XY does for its `secondary` argument but that does not indicate any ORM instructions being shared between these two contexts.
> 
> if you want X.ys to load in terms of X.xys use the [association proxy](https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html)

Thank you very much for your response @zzzeek 

I've tried the usage with `AssociationProxy` which works but there is only one downside for this in terms of performance:
When accessed to `ys_proxy` defined on `X`, it will fetch all `xys` in a single query but then will emit a single query per `Y` object that is associated to that `xy`
Any idea how this can be solved so that accessing `ys_proxy` will emit no more than 2 queries (one for fetching the `xys` and one to fetch all of their corresponding `Y` records (if it is possible with a single query it would be even better, which is what happens when `ys` relationship is accessed)

Changes that were made:
```python

class X(Base):
    __tablename__ = "x"
    id = Column(Integer, primary_key=True)
    ys: List["Y"] = relationship("Y", secondary="xy", uselist=True)
    xys: List["XY"] = relationship("XY", uselist=True)

    ys_proxy = association_proxy("xys", "y") # ADDED

class XY(SoftDeleteable, Base):
    __tablename__ = "xy"
    x_id = Column(Integer, ForeignKey(X.id), primary_key=True)
    y_id = Column(Integer, ForeignKey(Y.id), primary_key=True)

    y: "Y" = relationship("Y") # ADDED

print(some_x.ys_proxy)
# The line above prints: [Y(value=Value 1), Y(value=Value 3)]
# which is good as the deleted XY associated with Y(value=Value 2) is not part of the result
# However, in terms of performance, too many SQL statements are emitted:
# SELECT xy.deleted_at AS xy_deleted_at, xy.x_id AS xy_x_id, xy.y_id AS xy_y_id 
# FROM xy 
# WHERE 1 = xy.x_id AND xy.deleted_at IS NULL
# ---
# SELECT y.id AS y_id, y.value AS y_value 
# FROM y 
# WHERE y.id = 1
# ---
# SELECT y.id AS y_id, y.value AS y_value 
# FROM y 
# WHERE y.id = 3
```

Your help is much appreciated Mike!

**Reply by zzzeek** [reply, marked as answer]
you have many options here, the simplest is to just have XY.y use joined eager loading (or selectin loading):

```py
class XY(SoftDeleteable, Base):
    __tablename__ = "xy"
    x_id = Column(Integer, ForeignKey(X.id), primary_key=True)
    y_id = Column(Integer, ForeignKey(Y.id), primary_key=True)

    y: "Y" = relationship("Y", lazy="joined") # ADDED "lazy='joined'"

```

the query then becomes:

```
SELECT xy.deleted_at AS xy_deleted_at, xy.x_id AS xy_x_id, xy.y_id AS xy_y_id, y_1.id AS y_1_id, y_1.value AS y_1_value 
FROM xy LEFT OUTER JOIN y AS y_1 ON y_1.id = xy.y_id 
WHERE %(param_1)s = xy.x_id AND xy.deleted_at IS NULL
2023-08-17 09:46:16,493 INFO sqlalchemy.engine.Engine [generated in 0.00025s] {'param_1': 1}
[Y(value=Value 1), Y(value=Value 3)]

```

if we set this to instead use `selectin` loading:

```py
    y: "Y" = relationship("Y", lazy="selectin") # ADDED
```

then we have two queries, but there's no LEFT OUTER JOIN used, it's potentially more efficient:

```
SELECT xy.deleted_at AS xy_deleted_at, xy.x_id AS xy_x_id, xy.y_id AS xy_y_id 
FROM xy 
WHERE %(param_1)s = xy.x_id AND xy.deleted_at IS NULL
2023-08-17 09:48:02,522 INFO sqlalchemy.engine.Engine [generated in 0.00021s] {'param_1': 1}
2023-08-17 09:48:02,525 INFO sqlalchemy.engine.Engine SELECT y.id AS y_id, y.value AS y_value 
FROM y 
WHERE y.id IN (%(primary_keys_1)s, %(primary_keys_2)s)
2023-08-17 09:48:02,525 INFO sqlalchemy.engine.Engine [generated in 0.00034s] {'primary_keys_1': 1, 'primary_keys_2': 3}
[Y(value=Value 1), Y(value=Value 3)]

```



thanks for the easy-to-run test!

**Reply by zzzeek** [reply]
the LEFT OUTER JOIN can also be made to a more efficient INNER join, which is fine here since y is not nullable, using innerjoin=True:

```py
    y: "Y" = relationship("Y", lazy="joined", innerjoin=True) # ADDED
```

since XY.y is many-to-one this is likely the best option as it produces the most succinct query:

```
SELECT xy.deleted_at AS xy_deleted_at, xy.x_id AS xy_x_id, xy.y_id AS xy_y_id, y_1.id AS y_1_id, y_1.value AS y_1_value 
FROM xy JOIN y AS y_1 ON y_1.id = xy.y_id 
WHERE %(param_1)s = xy.x_id AND xy.deleted_at IS NULL
2023-08-17 09:50:02,320 INFO sqlalchemy.engine.Engine [generated in 0.00052s] {'param_1': 1}
[Y(value=Value 1), Y(value=Value 3)]

```


eager loading for 1.4 is described at https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html

**Reply by yanivps** [reply]
Working great! (Marking as answer)

Just out of curiosity, adding `lazy="joined"` or `lazy="selectin"` to the `Y` relationship will cause `Y` to be loaded even if someone is just interested in the `xys` relationship (assuming there was some extra data on this model which currently does not exist in this example)

is there any way to perform this `lazy="joined"` only when the `y` relationship is accessed via the association_proxy `ys_proxy` that we've created?

**Reply by zzzeek** [reply]
yes but you have to indicate that up front using loader options, depending on what you want to do, see https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#relationship-loading-with-loader-options

the many-to-one joinedload for an association object like this is nearly always what someone wants though and it's fairly cheap

**Reply by zzzeek** [reply]
taht is, you could disable the load on a per query basis also, if we take the script and do this:

     some_x = session.query(X).options(defaultload(X.xys).lazyload(XY.y)).first()

that would cancel out the "joinedload" for XY.y.     Also no, you can't currently send association proxy attributes into loader options, you have to spell it out as above.
